tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/LogsInjection/DiagnosticContextHelper.cs

```diff
@@ -0,0 +1,136 @@
+﻿// <copyright file="DiagnosticContextHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.NLog.LogsInjection
+{
+    internal static class DiagnosticContextHelper
+    {
+        public static MappedDiagnosticsLogicalContextSetterProxy GetMdlcProxy(Assembly nlogAssembly)
+        {
+            var mdclType = nlogAssembly.GetType("NLog.MappedDiagnosticsLogicalContext");
+            if (mdclType is not null)
+            {
+                // NLog 4.3+
+                var createTypeResult = DuckType.GetOrCreateProxyType(typeof(MappedDiagnosticsLogicalContextSetterProxy), mdclType);
```

Comment:
This type-checking looks good to me but I wonder if @tonyredondo has ideas on how to improve this. It'd be great to have a best practices guide of how to use the duck typing library for the best results 😄

Reply:
IIRC, I did it this way because some duck typing code paths don't work correctly with `struct`s (because requires boxing to get an `IDuckType`), and this keeps it consistent between `class` and `struct`, while using the same paths under the hood. That said, I wrote this originally before the 2.0 changes, so that may no longer be necessary!